daemon/api_notices_test.go

```diff
@@ -0,0 +1,285 @@
+	noticeId, err := st.AddNotice(state.ChangeUpdateNotice, "123", &state.AddNoticeOptions{
+		Data: map[string]string{"k": "v"},
+	})
+	c.Assert(err, IsNil)
+	st.Unlock()
+
+	query := makeQuery(after)
+	req, err := http.NewRequest("GET", "/v2/notices?"+query.Encode(), nil)
+	c.Assert(err, IsNil)
+	rsp := s.syncReq(c, req, nil)
+	c.Check(rsp.Status, Equals, 200)
+
+	notices, ok := rsp.Result.([]*state.Notice)
+	c.Assert(ok, Equals, true)
+	c.Assert(notices, HasLen, 1)
+	n := noticeToMap(c, notices[0])
+
+	firstOccurred, err := time.Parse(time.RFC3339, n["first-occurred"].(string))
+	c.Assert(err, IsNil)
+	c.Assert(firstOccurred.After(after), Equals, true)
+	lastOccurred, err := time.Parse(time.RFC3339, n["last-occurred"].(string))
+	c.Assert(err, IsNil)
+	c.Assert(lastOccurred.Equal(firstOccurred), Equals, true)
+	lastRepeated, err := time.Parse(time.RFC3339, n["last-repeated"].(string))
+	c.Assert(err, IsNil)
+	c.Assert(lastRepeated.Equal(firstOccurred), Equals, true)
+
+	delete(n, "first-occurred")
+	delete(n, "last-occurred")
+	delete(n, "last-repeated")
+	c.Assert(n, DeepEquals, map[string]any{
+		"id":           noticeId,
+		"type":         "change-update",
+		"key":          "123",
+		"occurrences":  1.0,
+		"last-data":    map[string]any{"k": "v"},
+		"expire-after": "168h0m0s",
+	})
+}
+
+func (s *noticesSuite) TestNoticesFilterMultipleTypes(c *C) {
+	s.daemon(c)
+
+	st := s.d.Overlord().State()
+	st.Lock()
+	addNotice(c, st, state.ChangeUpdateNotice, "123", nil)
+	time.Sleep(time.Microsecond)
+	addNotice(c, st, state.WarningNotice, "danger", nil)
+	st.Unlock()
+
+	req, err := http.NewRequest("GET", "/v2/notices?types=change-update&types=warning", nil)
+	c.Assert(err, IsNil)
+	rsp := s.syncReq(c, req, nil)
+	c.Check(rsp.Status, Equals, 200)
+
+	notices, ok := rsp.Result.([]*state.Notice)
+	c.Assert(ok, Equals, true)
+	c.Assert(notices, HasLen, 2)
+	n := noticeToMap(c, notices[0])
+	c.Assert(n["type"], Equals, "change-update")
+	n = noticeToMap(c, notices[1])
+	c.Assert(n["type"], Equals, "warning")
+}
+
+func (s *noticesSuite) TestNoticesFilterMultipleKeys(c *C) {
+	s.daemon(c)
+
+	st := s.d.Overlord().State()
+	st.Lock()
+	addNotice(c, st, state.ChangeUpdateNotice, "123", nil)
+	time.Sleep(time.Microsecond)
+	addNotice(c, st, state.ChangeUpdateNotice, "456", nil)
+	time.Sleep(time.Microsecond)
+	addNotice(c, st, state.WarningNotice, "danger", nil)
+	st.Unlock()
+
+	req, err := http.NewRequest("GET", "/v2/notices?keys=456&keys=danger", nil)
+	c.Assert(err, IsNil)
+	rsp := s.syncReq(c, req, nil)
+	c.Check(rsp.Status, Equals, 200)
+
+	notices, ok := rsp.Result.([]*state.Notice)
+	c.Assert(ok, Equals, true)
+	c.Assert(notices, HasLen, 2)
+	n := noticeToMap(c, notices[0])
+	c.Assert(n["key"], Equals, "456")
+	n = noticeToMap(c, notices[1])
+	c.Assert(n["key"], Equals, "danger")
+}
+
+func (s *noticesSuite) TestNoticesWait(c *C) {
+	s.daemon(c)
+
+	st := s.d.Overlord().State()
+	go func() {
+		time.Sleep(10 * time.Millisecond)
```

Comment:
If we do see any timing-related issues and end up having to mock WaitNotices, we could also change this one